harness/determined/common/api/_session.py

```diff
@@ -100,13 +158,102 @@ def put(
     ) -> requests.Response:
         return self._do_request("PUT", path, params, json, data, headers, timeout, False)
 
+
+class UnauthSession(BaseSession):
+    """
+    UnauthSession is mostly only useful to log in or unathenticated endpoints like /info.
+    """
+
+    def __init__(
+        self,
+        master: Optional[str],
+        cert: Optional[certs.Cert],
+        max_retries: Optional[urllib3.util.retry.Retry] = None,
+    ) -> None:
+        self.master = master or util.get_default_master_address()
+        self.cert = cert
+        self._max_retries = max_retries
+
+    def _do_request(
+        self,
+        method: str,
+        path: str,
+        params: Optional[Dict[str, Any]],
+        json: Any,
+        data: Optional[str],
+        headers: Optional[Dict[str, Any]],
+        timeout: Optional[int],
+        stream: bool,
+    ) -> requests.Response:
+        return _do_request(
+            method=method,
+            host=self.master,
+            path=path,
+            params=params,
+            json=json,
+            data=data,
+            headers=headers,
+            cert=self.cert,
+            timeout=timeout,
+            stream=stream,
+            max_retries=self._max_retries,
+        )
+
+
+class Session(BaseSession):
+    """
+    Session authenticates every request it makes.
+
+    By far, most BaseSessions in the codebase will be this Session subclass.
+    """
+
+    def __init__(
+        self,
+        master: Optional[str],
+        utp: authentication.UsernameTokenPair,
+        cert: Optional[certs.Cert],
+        max_retries: Optional[urllib3.util.retry.Retry] = None,
+    ) -> None:
+        self.master = master or util.get_default_master_address()
+        self.username = utp.username
+        self.token = utp.token
+        self.cert = cert
+        self._max_retries = max_retries
+
+    def _do_request(
+        self,
+        method: str,
+        path: str,
+        params: Optional[Dict[str, Any]],
+        json: Any,
+        data: Optional[str],
+        headers: Optional[Dict[str, Any]],
+        timeout: Optional[int],
+        stream: bool,
+    ) -> requests.Response:
+        # Add authentication.
+        headers = dict(headers) if headers is not None else {}
+        headers["Authorization"] = f"Bearer {self.token}"
+        return _do_request(
+            method,
+            self.master,
+            path,
+            params=params,
+            json=json,
+            data=data,
+            cert=self.cert,
+            headers=headers,
+            timeout=timeout,
+            stream=stream,
+            max_retries=self._max_retries,
+        )
+
     def with_retry(self, retry: Optional[urllib3.util.retry.Retry]) -> "Session":
```

Comment:
can this be removed? don't see why you wouldn't just pass retry directly into the constructor.

Reply:
looks like it can be.

Actually, nobody's using it anywhere anymore.

Reply:
Whatever we do, I'd like it possible to effectively change the `Retry` attached to a `Session`. We're talking about making `login`s return a Session. I think that makes sense to do. But it makes less sense for `login` to take a `Retry` and for that `Retry` to be the same `Retry` used by the returned Session.